app/data/views.py

```diff
@@ -331,57 +332,152 @@ def _query_param_to_case_stmnt(self, param, request, queryset):
         }
 
-        # TODO: This sets year range based on the oldest and youngest records in the filtered
-        # queryset, but it (and possibly other date ranges) should be defined by the request's
-        # occurred_max and occurred_min parameters if available.
-        if period_type == 'year':
-            range_min = queryset.order_by('occurred_from').first().occurred_from.year
-            range_max = queryset.order_by('-occurred_from').first().occurred_from.year + 1
-            period_range = xrange(range_min, range_max)
-            period_labels = lambda x: str(x)
-        elif period_type in easy_ranges.keys():
-            period_range = easy_ranges[period_type]['keys']
-            period_labels = easy_ranges[period_type]['labels']
+        sequential_ranges = {
+            'year': {
+                'range': [],
+                'lookup': lambda x: {'occurred_from__year': x},
+                'label': lambda x: str(x)
+            },
+            'month': {
+                'range': [],
+                'lookup': lambda (yr, mo): {'occurred_from__month': mo, 'occurred_from__year': yr},
+                'label': lambda (yr, mo): '{}, {}'.format(calendar.month_name[mo], str(yr))
+            },
+            'week': {
+                'range': [],
+                'lookup': lambda (yr, wk): {'occurred_from__week': wk, 'occurred_from__year': yr},
+                'label': lambda (yr, wk): '{} Week {}'.format(str(yr), str(wk))
+            },
+            'day': {
+                'range': [],
+                'lookup': lambda (yr, wk, day): {'occurred_from__week': wk,
+                                                 'occurred_from__year': yr,
+                                                 'occurred_from__day': day},
+                'label': lambda (yr, wk, day): template_date(datetime.date(yr, wk, day))
+            },
+        }
+
+        if period_type in easy_ranges.keys():
+            period = easy_ranges[period_type]
+        elif period_type in sequential_ranges.keys():
+            # Get the desired range, either from the query params or the filtered queryset
+            if request.query_params.get('occurred_min') is not None:
+                min_date = parse_date(request.query_params['occurred_min']).date()
+            else:
+                min_date = queryset.order_by('occurred_from').first().occurred_from.date()
+            if request.query_params.get('occurred_max') is not None:
+                max_date = parse_date(request.query_params['occurred_max']).date()
+            else:
+                max_date = queryset.order_by('-occurred_from').first().occurred_from.date()
+
+            # Build the relevant range of aggregation periods. 'month' and 'week' depend on'year'
+            # and 'day' depends on 'month'
+            sequential_ranges['year']['range'] = xrange(min_date.year, max_date.year + 1)
+            if period_type != 'year':
+                sequential_ranges['month']['range'] = [
```

Comment:
These are mighty list comprehensions indeed, but I think they're readable and straightforward to think about.